File: controllers/rider.controller.js
const User = require('../models/User');
const Order = require('../models/Order');

// GET /api/riders/:id
exports.getRiderById = async (req, res) => {
  try {
    const rider = await User.findOne({ _id: req.params.id, role: 'rider' });
    if (!rider) return res.status(404).json({ message: 'Rider not found.' });
    res.json(rider);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// PUT /api/riders/:id
exports.updateRider = async (req, res) => {
  try {
    if (req.user.role !== 'admin' && req.user.id !== req.params.id) {
      return res.status(403).json({ message: 'Access denied.' });
    }

    const rider = await User.findOneAndUpdate(
      { _id: req.params.id, role: 'rider' },
      req.body,
      { new: true }
    );
    if (!rider) return res.status(404).json({ message: 'Rider not found.' });
    res.json(rider);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// PATCH /api/riders/:id/location
// Body contains { lat, lng, orderId }
exports.updateRiderLocation = async (req, res) => {
  try {
    if (req.user.role !== 'admin' && req.user.id !== req.params.id) {
      return res.status(403).json({ message: 'Access denied.' });
    }

    const { lat, lng, orderId } = req.body;
    if (lat === undefined || lng === undefined) {
      return res.status(400).json({ message: 'Latitude and longitude are required.' });
    }

    // Save location to rider's profile
    const rider = await User.findOneAndUpdate(
      { _id: req.params.id, role: 'rider' },
      { 'location.lat': lat, 'location.lng': lng },
      { new: true }
    );

    if (!rider) return res.status(404).json({ message: 'Rider not found.' });

    // Emit live location updates via Socket.io
    const io = req.app.get('io');
    if (io) {
      // 1. Broadcast to the order room
      if (orderId) {
        io.to(`order_${orderId}`).emit('rider:locationUpdate', { riderId: req.params.id, lat, lng });
      }
      // 2. Also broadcast to customer room directly if we have the order
      if (orderId) {
        const order = await Order.findById(orderId);
        if (order) {
          io.to(`user_${order.customer}`).emit('rider:locationUpdate', { orderId, lat, lng });
        }
      }
    }

    res.json({ message: 'Location updated successfully.', location: { lat, lng } });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// GET /api/riders/:id/deliveries
exports.getRiderDeliveries = async (req, res) => {
  try {
    if (req.user.role !== 'admin' && req.user.id !== req.params.id) {
      return res.status(403).json({ message: 'Access denied.' });
    }

    const deliveries = await Order.find({ rider: req.params.id })
      .populate('customer', 'name email phone location')
      .populate('vendor', 'name address phone location')
      .sort('-createdAt');

    res.json(deliveries);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

// PATCH /api/riders/:id/availability
// Body contains { availability: true | false }
exports.updateRiderAvailability = async (req, res) => {
  try {
    if (req.user.role !== 'admin' && req.user.id !== req.params.id) {
      return res.status(403).json({ message: 'Access denied.' });
    }

    const { availability } = req.body;
    if (availability === undefined) {
      return res.status(400).json({ message: 'Availability boolean value is required.' });
    }

    const rider = await User.findOneAndUpdate(
      { _id: req.params.id, role: 'rider' },
      { availability },
      { new: true }
    );

    if (!rider) return res.status(404).json({ message: 'Rider not found.' });
    res.json(rider);
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

File: controllers/rider_controller.py
from fastapi import HTTPException, status
from database import SyncSessionLocal, serialize_doc
from schemas.rider import RiderUpdate, LocationSchema, AvailabilitySchema
from models.rider import Rider
from models.order import Order, OrderItem

async def get_rider_by_id(rider_id: int, current_user: dict):
    session = SyncSessionLocal()
    try:
        rider = session.query(Rider).filter(Rider.id == rider_id).first()
        if not rider:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Rider not found")

        serialized = serialize_doc(rider)
        serialized.pop("password", None)
        return serialized
    finally:
        session.close()

async def update_rider(rider_id: int, data: RiderUpdate, current_user: dict):
    if current_user.get("role") != "admin" and current_user.get("id") != rider_id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Unauthorized to update rider profile")

    session = SyncSessionLocal()
    try:
        rider = session.query(Rider).filter(Rider.id == rider_id).first()
        if not rider:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Rider not found")

        update_dict = {k: v for k, v in data.dict(exclude_unset=True).items() if v is not None}
        for k, v in update_dict.items():
            if hasattr(rider, k):
                setattr(rider, k, v)

        session.commit()
        session.refresh(rider)

        serialized = serialize_doc(rider)
        serialized.pop("password", None)
        return serialized
    finally:
        session.close()

async def update_rider_location(rider_id: int, data: LocationSchema, current_user: dict):
    if current_user.get("role") != "admin" and current_user.get("id") != rider_id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Unauthorized to update location")

    session = SyncSessionLocal()
    try:
        rider = session.query(Rider).filter(Rider.id == rider_id).first()
        if not rider:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Rider not found")

        rider.latitude = data.latitude
        rider.longitude = data.longitude
        session.commit()

        return {
            "message": "Location updated successfully",
            "latitude": data.latitude,
            "longitude": data.longitude
        }
    finally:
        session.close()

async def update_rider_availability(rider_id: int, data: AvailabilitySchema, current_user: dict):
    if current_user.get("role") != "admin" and current_user.get("id") != rider_id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Unauthorized to update availability")

    session = SyncSessionLocal()
    try:
        rider = session.query(Rider).filter(Rider.id == rider_id).first()
        if not rider:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Rider not found")

        rider.is_available = data.is_available
        session.commit()
        session.refresh(rider)

        serialized = serialize_doc(rider)
        serialized.pop("password", None)
        return serialized
    finally:
        session.close()

async def get_rider_deliveries(rider_id: int, current_user: dict):
    if current_user.get("role") != "admin" and current_user.get("id") != rider_id:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Unauthorized to access rider deliveries")

    session = SyncSessionLocal()
    try:
        deliveries = session.query(Order).filter(Order.rider_id == rider_id).order_by(Order.created_at.desc()).all()
        serialized_list = serialize_doc(deliveries)
        for d in serialized_list:
            items = session.query(OrderItem).filter(OrderItem.order_id == d["id"]).all()
            d["items"] = serialize_doc(items)
        return serialized_list
    finally:
        session.close()
